index.js

```diff
@@ -4413,6 +4413,53 @@
     impl: matchAll
   };
 
+  //# replace :: (Array (Maybe String) -> String) -> RegExp -> String -> String
+  //.
+  //. Replaces occurrences of the given pattern within the given string
+  //. in accordance with the given replacement function, which receives an
+  //. array of captured values. Replaces all occurrences of the pattern if
+  //. its `g` flag is set; just the first occurrence otherwise.
+  //.
+  //. ```javascript
+  //. > S.replace (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/) ('foo')
+  //. 'Foo'
+  //.
+  //. > S.replace (([$1]) => S.maybe ('') (S.toUpper) ($1)) (/(\w)/g) ('foo')
+  //. 'FOO'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<>')
+  //. '<>'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<foo>')
+  //. '<[Just ("foo"), Nothing]>'
+  //.
+  //. > S.replace (S.show) (/(foo)(bar)?/) ('<foobar>')
+  //. '<[Just ("foo"), Just ("bar")]>'
+  //. ```
+  function replace(substitute) {
+    return function(pattern) {
+      return function(text) {
+        return text.replace (pattern, function() {
+          var groups = [];
```

Comment:
Is it possible to know the length of this array upfront? Say, using `var groups = new Array(arguments.length)`. If so, then that would get you slightly better performance at relatively low implementation and maintenance cost. You would have to alter `groups.push` statements to index assignment statements instead, eg: `groups[idx] = (...)`.

Reply:
> Is it possible to know the length of this array upfront?

No, due to the existence of named capturing groups:

```javascript
> '@sanctuary-js'.replace (/@([-\w]+)/, function() { return arguments.length; })
'4'

> '@sanctuary-js'.replace (/@(?<username>[-\w]+)/, function() { return arguments.length; })
'5'
```

Reply:
Ah, so these named groups are skipped? Is that what the `typeof` check below is doing?

Reply:
That's right. I imagine `groups` is omitted when the pattern contains no named capturing groups to preserve backward compatibility. Providing the sensible empty value, `{}`, as an additional argument would break programs that assume that `Array.prototype.slice.call(arguments, 1, -2)` extracts (only) the captured values.

Reply:
Oh, wow. It took me a really long time to realize that what's happening there is: *skip one argument, and keep taking arguments until a number is encountered*. The numeric argument signifies the "end" of the capture group arguments. Wow. The signature of JavaScript's `replace` is crazier than I remember.